text/0003-templates.md

```diff
@@ -0,0 +1,258 @@
+
+### Toolkit
+
+The `pnpm.templates.toolkit` field copies the `dependencies` block of the template and spreads it into the `devDependencies` of the referencing package.
+
+**On-Disk**
+
+```json5
+{
+  "name": "@example/react-components",
+  "pnpm": {
+    "templates": {
+      "toolkit": ["@example/react-template@0.1.0"],
+    }
+  },
+  "devDependencies": {
+    "jest": "^29.4.3"
+  }
+}
+```
+
+**In-Memory and Publish Time**
+
+```json5
+{
+  "name": "@example/react-components",
+  "dependencies": {
+    "redux": "^4.2.1"
+  },
+  "devDependencies": {
+    "jest": "^29.4.3",
+    "react": "^18.2.0",
+    "react-dom": "^18.2.0",
+  }
+}
+```
+
+This allows package authors to share a reusable set of tools that can be imported or required in the referencing package. Dependencies of dependencies are not normally visible in this manner since pnpm sets up a semi-strict `node_modules` structure by default.
+
+### Catalog
+
+The `pnpm.templates.catalog` feature allows packages to declare a dependency using a version specifier from the referenced template. The `dependencies` block of the template will be available to reference through the `catalog:` version specifier protocol.
+
+**On-Disk**
+
+```json5
+{
+  "name": "@example/react-components",
+  "pnpm": {
+    "templates": {
+      "catalog": ["@example/react-template@0.1.0"],
+    }
+  },
+  "dependencies": {
+    "react": "catalog:"
+  }
+}
+```
+
+**In-Memory and Publish Time**
+
+```json5
+{
+  "name": "@example/react-components",
+  "dependencies": {
+    "react": "^18.2.0"
+  }
+}
+```
+
+We expect most monorepos to use this feature to keep dependency specifiers consistent between different in-repo packages.
+
+### Combining Templates
+
+Templates may not refer to other templates. Instead of an inheritance hierarchy, a package may refer to multiple templates with later entries in the list taking precedence.
+
+For example, if both `@organization/authoring-metadata` and `@team/authoring-metadata` have an `author` field, the value from `@team/authoring-metadata` will be used.
+
+```json5
+{
+  "name": "@example/simple",
+  "pnpm": {
+    "templates": {
+      "extends": [
+        "@organization/authoring-metadata@0.1.0",
+        "@team/authoring-metadata@0.1.0"
+      ],
+    }
+  }
+}
+```
+
+Avoiding inheritance simplifies implementation and usability in several ways.
+
+- Suppose a new version of pnpm adds `pnpm.templates` config options. In an inheritance model, using these new options on a template would force consumers to a greater minimum version of pnpm.
+- Complex circular resolution during installation is avoided.
+- Template loading is more performant. The requirement to declare all required templates up front enables fetching in a single parallelized step, rather than a waterfall fetch at each level of a theoretical template inheritance hierarchy.
```

Comment:
I don't think these are issues. We will only allow to reference templates by exact versions. So when a template inherits another template, we will actually merge the parent template into the current one and publish the current template as a single independent template.

Reply:
Agree most of the problems in this list go away if the parent template is rendered/merged into the current one before publishing.

- In that world, should we still allow single inheritance only, or introduce multiple inheritance to make it easier to reference templates from different authors? I was following the [TypeScript 5.0 changes and noticed `tsconfig.json` files will allow multiple `extends` entries](https://devblogs.microsoft.com/typescript/announcing-typescript-5-0-beta/#supporting-multiple-configuration-files-in-extends).
- The `toolkit` templates are difficult to model with inheritance since the render step moves `dependencies` to `devDependencies`. It's possible we don't want that to happen in a template. But that introduces complexity since the render/merge step now depends on whether the current package is itself a template.

Reply:
I don't have objections to support multiple inheritance.

Reply:
> We will only allow to reference templates by exact versions.

I would likely want to use this as `my-template@workspace:*` instead of a specific version, within a workspace.

Reply:
I had imagined templates being usable through `my-template@workspace:*` as well. Will explicitly mention that in the RFC after a few more discussions resolve.